menas/src/main/scala/za/co/absa/enceladus/menas/services/DatasetService.scala

```diff
@@ -102,4 +101,147 @@ class DatasetService @Autowired() (datasetMongoRepository: DatasetMongoRepositor
     }
   }
 
+  override def importItem(item: Dataset, username: String): Future[Option[Dataset]] = {
+    getLatestVersionValue(item.name).flatMap {
+      case Some(version) => update(username, item.copy(version = version))
+      case None => super.create(item.copy(version = 1), username)
+    }
+  }
+
+  override def validateSingleImport(item: Dataset, metadata: Map[String, String]): Future[Validation] = {
+    val confRulesWithConnectedEntities = item.conformance.filter(_.hasConnectedEntities)
+    val maybeSchema = datasetMongoRepository.getConnectedSchema(item.schemaName, item.schemaVersion)
+
+    val validations = super.validateSingleImport(item, metadata)
+    val validationsWithSchema = validateSchema(item.schemaName, item.schemaVersion, validations, maybeSchema)
+    val validationWithConnected = validateConnectedEntitiesExistence(confRulesWithConnectedEntities, validationsWithSchema)
+    validateConformanceRules(item.conformance, maybeSchema, validationWithConnected)
+  }
+
+  private def validateConnectedEntitiesExistence(confRulesWithConnectedEntities: List[ConformanceRule],
+                                                 validations: Future[Validation]): Future[Validation] = {
+    def standardizedErrMessage(ce: ConnectedEntity) = s"Connected ${ce.kind} ${ce.name} v${ce.version} could not be found"
+
+    confRulesWithConnectedEntities.foldLeft(validations) { (acc, cr) =>
+      val connectedEntities = cr.connectedEntities.map {
+        case mt: ConnectedMappingTable => (mt, datasetMongoRepository.getConnectedMappingTable(mt.name, mt.version))
+      }
+
+      connectedEntities.foldLeft(acc) { case (v, (entityDef, entityDBInstance)) =>
+        for {
+          instance <- entityDBInstance
+          accValidations <- v
+        } yield accValidations.withErrorIf(instance.isEmpty, s"item.${entityDef.kind}", standardizedErrMessage(entityDef))
```

Comment:
Seems like I don't have enough deep understanding on for comprehension. 
Why would this not work?
![image](https://user-images.githubusercontent.com/17467768/97208845-4135b600-17c4-11eb-9ec5-1bf524a924c3.png)

Reply:
I am generally confused by this part of code. Another topic for discussion, please?

Reply:
I think you cannot directly mix-and-match different kind of functors in a single for comprehension. You might need a monad transformer to do this. I actually find [this article on the topic](https://blog.buildo.io/monad-transformers-for-the-working-programmer-aa7e981190e7) both equally entertaining and educative.